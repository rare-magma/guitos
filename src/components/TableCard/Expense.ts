import { ItemForm } from "../ItemForm/ItemForm";

export class Expense {
  items: ItemForm[] = [];
  total = 0;

  constructor(initializer?: Expense) {
    if (!initializer) return;
    if (initializer.items) this.items = initializer.items;
    if (initializer.total) this.total = initializer.total;
  }
}
