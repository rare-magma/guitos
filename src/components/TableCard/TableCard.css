.btn-Expenses-plus-button {
  border: 1px solid var(--comment);
  color: var(--purple);
}

.btn-Expenses-plus-button:focus-visible,
.btn-Expenses-plus-button:hover {
  background-color: var(--purple);
  border: 1px solid var(--purple);
  color: var(--lightbgcolor);
}

.btn-Revenue-plus-button {
  border: 1px solid var(--comment);
  color: var(--highlight);
}

.btn-Revenue-plus-button:focus-visible,
.btn-Revenue-plus-button:hover {
  background-color: var(--highlight);
  border: 1px solid var(--highlight);
  color: var(--lightbgcolor);
}

.Expenses-card,
.Revenue-card {
  border: 1px solid var(--comment);
}

.Expenses-card-header,
.Revenue-card-header {
  border-bottom: 1px solid var(--comment);
}
