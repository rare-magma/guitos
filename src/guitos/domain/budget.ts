import { immerable } from "immer";
import type { BudgetItem } from "./budgetItem";
import type { Expenses } from "./expenses";
import type { Incomes } from "./incomes";
import type { Stats } from "./stats";
import { Uuid } from "./uuid";

export class Budget {
  id: Uuid;
  name: string;
  expenses: Expenses;
  incomes: Incomes;
  stats: Stats;

  [immerable] = true;

  constructor(
    id: Uuid,
    name: string,
    expenses: Expenses,
    incomes: Incomes,
    stats: Stats,
  ) {
    this.id = id;
    this.name = name;
    this.expenses = expenses;
    this.incomes = incomes;
    this.stats = stats;
  }

  static create(date?: string, goal?: number): Budget {
    const newId = Uuid.random();
    const year = new Date().getFullYear();
    const newBudget = new Budget(
      newId,
      date ?? `${year}-${newId.toString().slice(0, 8)}`,
      {
        items: [{ id: 1, name: "", value: 0 }],
        total: 0,
      },
      {
        items: [{ id: 1, name: "", value: 0 }],
        total: 0,
      },
      {
        available: 0,
        withGoal: 0,
        saved: 0,
        goal: goal ?? 10,
        reserves: 0,
      },
    );

    return newBudget;
  }

  static createEmpty(name: string): Budget {
    const emptyExpenses: BudgetItem[] = [];
    const emptyIncomes: BudgetItem[] = [];
    const emptyBudget: Budget = {
      ...Budget.create(name, 0),
      expenses: {
        items: emptyExpenses,
        total: 0,
      },
      incomes: {
        items: emptyIncomes,
        total: 0,
      },
    };
    return emptyBudget;
  }

  static clone(budget: Budget): Budget {
    const clonedBudget: Budget = {
      ...budget,
      id: Uuid.random(),
      name: `${budget.name}-clone`,
    };
    return clonedBudget;
  }

  static toSafeFormat(budget: Budget) {
    return {
      ...budget,
      id: budget.id.toString(),
    };
  }
}
