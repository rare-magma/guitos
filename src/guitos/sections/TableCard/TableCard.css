.btn-expenses-plus-button {
  background-color: var(--bgcolor);
  border: 1px solid var(--lightbgcolor);
  color: var(--purple);
}

.btn-expenses-plus-button:focus-visible,
.btn-expenses-plus-button:hover {
  background-color: var(--purple);
  border: 1px solid var(--purple);
  color: var(--lightbgcolor);
}

.btn-revenue-plus-button {
  background-color: var(--bgcolor);
  border: 1px solid var(--lightbgcolor);
  color: var(--highlight);
}

.btn-revenue-plus-button:focus-visible,
.btn-revenue-plus-button:hover {
  background-color: var(--highlight);
  border: 1px solid var(--highlight);
  color: var(--lightbgcolor);
}

.expenses-card,
.revenue-card {
  border: 1px solid var(--lightbgcolor);
}

.expenses-card-header,
.revenue-card-header {
  border-bottom: 1px solid var(--comment);
}

.btn-check:checked + .btn,
.btn.active,
.btn.show {
  box-shadow: none !important;
  background-color: var(--cyan);
  border: 1px solid var(--cyan);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.btn-check + .btn:hover {
  box-shadow: none !important;
  border: 1px solid var(--cyan);
}

.btn-check:focus-visible + .btn,
.btn-check {
  box-shadow: none !important;
  color: var(--cyan);
  border: 1px solid var(--cyan);
}

.btn-outline-info.toggle {
  background-color: var(--bgcolor);
  border: 1px solid var(--lightbgcolor);
  border-radius: 0.375rem;
  color: var(--cyan);
}

.btn-outline-info.toggle:focus-visible,
.btn-outline-info.toggle:hover {
  background-color: var(--cyan);
  border: 1px solid var(--cyan);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.btn-outline-info.toggle:disabled {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--lightbgcolor);
  border-radius: 0.375rem;
  color: var(--comment);
}
