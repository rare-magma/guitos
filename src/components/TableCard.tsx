import { useState } from "react";
import { ItemForm } from "./ItemForm";
import {
  Card,
  Button,
  Row,
  Col,
  OverlayTrigger,
  Tooltip,
} from "react-bootstrap";
import ItemFormGroup from "./ItemFormGroup";
import {
  calcPercentage,
  calcTotal,
  intlFormat,
  round,
  userLang,
} from "../utils";
import { Expense } from "./Expense";
import { Income } from "./Income";
import { BsPlusLg } from "react-icons/bs";

interface TableCardProps {
  items: Income | Expense;
  revenueTotal: number;
  header: string;
  onChange: (table: Income | Expense) => void;
}

function TableCard({
  items: initialItems,
  revenueTotal,
  header: label,
  onChange,
}: TableCardProps) {
  const [table, setTable] = useState(initialItems);
  const [total, setTotal] = useState(calcTotal(table.items));
  const revenuePercentage = calcPercentage(total, revenueTotal);

  const addTable = (tableBeingEdited: Income | Expense) => {
    let newTable;
    if (label === "Revenue") {
      newTable = new Income();
    } else if (label === "Expenses") {
      newTable = new Expense();
    } else {
      throw new Error("Messed up table type");
    }

    const newItemForm = new ItemForm();

    let maxId;
    if (table.items.length !== 0) {
      maxId = Math.max(
        ...table.items.map((i) => {
          return i.id;
        })
      );
    } else {
      maxId = 0;
    }

    newItemForm.id = isNaN(maxId) ? 0 : maxId + 1;
    newItemForm.name = "";
    newItemForm.value = 0;

    newTable.items = tableBeingEdited.items.concat(newItemForm);
    newTable.total = calcTotal(newTable.items);

    setTable(newTable);
    setTotal(calcTotal(newTable.items));
    onChange(newTable);
  };

  const removeTable = (toBeDeleted: ItemForm) => {
    let newTable;
    if (toBeDeleted.constructor.name === "Income") {
      newTable = new Income();
    } else {
      newTable = new Expense();
    }
    newTable.items = table.items.filter(
      (item: { id: number }) => item.id !== toBeDeleted.id
    );
    newTable.total = calcTotal(newTable.items);

    setTable(newTable);
    setTotal(calcTotal(newTable.items));
    onChange(newTable);
  };

  const handleChange = (item: ItemForm) => {
    let newTable;
    if (item.constructor.name === "Income") {
      newTable = new Income();
    } else {
      newTable = new Expense();
    }
    newTable.items = table.items.map((i) => {
      if (i.id === item.id) {
        return {
          id: item.id,
          name: item.name,
          value: isNaN(item.value) ? 0 : Number(item.value),
        };
      }
      return i;
    });
    newTable.total = calcTotal(newTable.items);

    setTable(newTable);
    setTotal(newTable.total);
    onChange(newTable);
  };

  return (
    <Card className={label + "-card"}>
      <Card.Header className={label + "-card-header"}>
        <Row>
          <Col>{label}</Col>
          <Col className="text-end">
            <OverlayTrigger
              placement="top"
              overlay={
                label === "Expenses" ? (
                  <Tooltip id={`tooltip-value-${label}`}>
                    {revenuePercentage}% of revenue
                  </Tooltip>
                ) : (
                  <></>
                )
              }
            >
              <div>{intlFormat(round(total * 100, 2), userLang)}</div>
            </OverlayTrigger>
          </Col>
        </Row>
      </Card.Header>
      <Card.Body>
        {table.items?.map((item: ItemForm) => (
          <ItemFormGroup
            key={item.id}
            itemForm={item}
            costPercentage={calcPercentage(item.value, revenueTotal)}
            onChange={handleChange}
            onRemove={() => {
              removeTable(item);
            }}
          />
        ))}
        <div className="mt-3" />
        <div className="d-grid gap-2">
          <Button
            variant={label + "-plus-button"}
            size="sm"
            onClick={() => {
              addTable(table);
            }}
            name=""
            value=""
          >
            <BsPlusLg />
          </Button>
        </div>
      </Card.Body>
    </Card>
  );
}

export default TableCard;
