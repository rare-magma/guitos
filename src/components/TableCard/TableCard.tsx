import Big from "big.js";
import { Reorder } from "framer-motion";
import { produce } from "immer";
import { useRef, useState } from "react";
import {
  Button,
  Card,
  Col,
  OverlayTrigger,
  Row,
  ToggleButton,
  Tooltip,
} from "react-bootstrap";
import { BsArrowsVertical, BsPlusLg } from "react-icons/bs";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import {
  calcAvailable,
  calcPercentage,
  calcSaved,
  calcTotal,
  calcWithGoal,
  intlFormat,
  roundBig,
} from "../../utils";
import { ItemForm } from "../ItemForm/ItemForm";
import { ItemFormGroup } from "../ItemForm/ItemFormGroup";
import { Expense } from "./Expense";
import { Income } from "./Income";
import "./TableCard.css";

interface TableCardProps {
  header: "Revenue" | "Expenses";
}

export function TableCard({ header: label }: TableCardProps) {
  const { budget, setBudget, revenuePercentage } = useBudget();
  const { intlConfig } = useConfig();
  const inputRef = useRef<HTMLInputElement>(null);
  const [isDraggable, setIsDraggable] = useState(false);

  const isRevenue = label === "Revenue";
  const isExpense = label === "Expenses";
  const table = isExpense ? budget?.expenses : budget?.incomes;
  const total = table?.total ?? 0;

  function reorderTable(newOrder: ItemForm[]) {
    if (!budget) return;
    const newState = produce((draft) => {
      isExpense
        ? (draft.expenses.items = newOrder)
        : (draft.incomes.items = newOrder);
    }, budget);
    setBudget(newState(), true);
  }

  function handleTableChange(item: Income | Expense) {
    if (!budget) return;
    const newState = produce((draft) => {
      isExpense ? (draft.expenses = item) : (draft.incomes = item);
      draft.stats.available = roundBig(calcAvailable(draft), 2);
      draft.stats.withGoal = calcWithGoal(draft);
      draft.stats.saved = calcSaved(draft);
    }, budget);
    setBudget(newState(), true);
  }

  function addItemToTable(tableBeingEdited: Income | Expense | undefined) {
    if (!tableBeingEdited) return;
    const tableHasItems = table && table.items.length !== 0;
    const newItemForm = {} as ItemForm;
    const newTable = isRevenue ? ({} as Income) : ({} as Expense);
    let maxId: number;

    if (tableHasItems) {
      maxId = Math.max(
        ...table.items.map((i) => {
          return i.id;
        }),
      );
    } else {
      maxId = 0;
    }

    newItemForm.id = isNaN(maxId) ? 0 : maxId + 1;
    newItemForm.name = "";
    newItemForm.value = 0;

    newTable.items = tableBeingEdited.items.concat(newItemForm);
    newTable.total = roundBig(calcTotal(newTable.items), 2);

    handleTableChange(newTable);
  }

  return (
    <Card
      key={`table-${label}-${intlConfig?.currency}-${table?.items.length}`}
      className={label + "-card"}
    >
      <Card.Header className={label + "-card-header"}>
        <OverlayTrigger
          placement="top"
          overlay={
            isExpense ? (
              <Tooltip
                id={`tooltip-value-${label}`}
                style={{ position: "fixed" }}
              >
                {revenuePercentage}% of revenue
              </Tooltip>
            ) : (
              <></>
            )
          }
        >
          <Row>
            <Col>{label}</Col>
            <Col className="text-end fixed-width-font">
              <div aria-label="total amount">
                {intlConfig?.currency &&
                  intlFormat(roundBig(Big(total), 2), intlConfig?.currency)}
              </div>
            </Col>
          </Row>
        </OverlayTrigger>
      </Card.Header>
      <Card.Body>
        {table?.items && (
          <Reorder.Group
            axis="y"
            values={table.items}
            onReorder={reorderTable}
            as="div"
          >
            {table.items?.map((item: ItemForm) => (
              <Reorder.Item
                key={item.id}
                value={item}
                as="div"
                dragListener={isDraggable}
              >
                <ItemFormGroup
                  key={`${label}-${item.id}-item-form-group`}
                  itemForm={item}
                  label={label}
                  costPercentage={
                    budget
                      ? calcPercentage(item.value, budget.incomes.total)
                      : 0
                  }
                  inputRef={inputRef}
                />
              </Reorder.Item>
            ))}
          </Reorder.Group>
        )}
        <div className="mt-3" />
        <div className="d-flex gap-2">
          <OverlayTrigger
            delay={250}
            placement="top"
            overlay={
              <Tooltip
                id={`tooltip-new-itemformgroup`}
                style={{ position: "fixed" }}
              >
                add new item
              </Tooltip>
            }
          >
            <Button
              variant={label.toLowerCase() + "-plus-button"}
              aria-label={`add item to ${label}`}
              size="sm"
              className="flex-grow-1"
              onClick={() => {
                addItemToTable(table);
                setTimeout(() => {
                  inputRef.current?.focus();
                }, 0);
              }}
              name=""
              value=""
            >
              <BsPlusLg aria-hidden />
            </Button>
          </OverlayTrigger>
          <OverlayTrigger
            delay={250}
            placement="top"
            overlay={
              <Tooltip
                id={`tooltip-reorder-itemformgroup`}
                style={{ position: "fixed" }}
              >
                toggle reordering of items
              </Tooltip>
            }
          >
            <ToggleButton
              id={`toggle-reorder-${label}`}
              aria-label={`reorder items in ${label}`}
              type="checkbox"
              variant="outline-info"
              value={1}
              size="sm"
              onClick={() => setIsDraggable(!isDraggable)}
              checked={isDraggable}
            >
              <BsArrowsVertical aria-hidden />
            </ToggleButton>
          </OverlayTrigger>
        </div>
      </Card.Body>
    </Card>
  );
}

export default TableCard;
