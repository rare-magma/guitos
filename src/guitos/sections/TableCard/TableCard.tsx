import Big from "big.js";
import { produce } from "immer";
import { Reorder } from "motion/react";
import { useRef, useState } from "react";
import {
  Button,
  Card,
  Col,
  OverlayTrigger,
  Row,
  ToggleButton,
  Tooltip,
} from "react-bootstrap";
import { BsArrowsVertical, BsPlusLg } from "react-icons/bs";
import { intlFormat, roundBig } from "../../../utils";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { ItemFormGroup } from "../ItemForm/ItemFormGroup";
import "./TableCard.css";
import { BudgetCalculator } from "../../application/budgetCalculator";
import type { Budget } from "../../domain/budget";
import { BudgetItem } from "../../domain/budgetItem";
import type { Expenses } from "../../domain/expenses";
import type { Incomes } from "../../domain/incomes";

interface TableCardProps {
  header: "Revenue" | "Expenses";
}

// biome-ignore lint/style/noDefaultExport: lazy loading
export default function TableCard({ header: label }: TableCardProps) {
  const { budget, setBudget, revenuePercentage } = useBudget();
  const { userOptions } = useConfig();
  const inputRef = useRef<HTMLInputElement>(null);
  const [isDraggable, setIsDraggable] = useState(false);

  const isRevenue = label === "Revenue";
  const isExpense = label === "Expenses";
  const table = isExpense ? budget?.expenses : budget?.incomes;
  const total = table?.total ?? 0;

  function reorderTable(newOrder: BudgetItem[]) {
    if (!budget) return;
    const newState = produce((draft) => {
      if (isExpense) {
        draft.expenses.items = newOrder;
      } else {
        draft.incomes.items = newOrder;
      }
    }, budget);
    setBudget(newState(), true);
  }

  function handleTableChange(item: Incomes | Expenses) {
    if (!budget) return;
    const newState = produce((draft) => {
      if (isExpense) {
        draft.expenses = item;
      } else {
        draft.incomes = item;
      }
      draft.stats.available = roundBig(
        BudgetCalculator.available(draft as Budget),
        2,
      );
      draft.stats.withGoal = BudgetCalculator.availableWithGoal(
        draft as Budget,
      );
      draft.stats.saved = BudgetCalculator.saved(draft as Budget);
    }, budget);
    setBudget(newState(), true);
  }

  function addItemToTable(tableBeingEdited: Incomes | Expenses | undefined) {
    if (!tableBeingEdited) return;
    const tableHasItems = table && table.items.length !== 0;
    const newItemForm = BudgetItem.create();
    const newTable = isRevenue ? ({} as Incomes) : ({} as Expenses);
    let maxId: number;

    if (tableHasItems) {
      maxId = Math.max(
        ...table.items.map((i: BudgetItem) => {
          return i.id;
        }),
      );
    } else {
      maxId = 0;
    }

    newItemForm.id = Number.isNaN(maxId) ? 0 : maxId + 1;
    newItemForm.name = "";
    newItemForm.value = 0;

    newTable.items = tableBeingEdited.items.concat(newItemForm);
    newTable.total = roundBig(BudgetCalculator.itemsTotal(newTable.items), 2);

    handleTableChange(newTable);
  }

  return (
    <Card
      key={`table-${label}-${userOptions.currencyCode}-${table?.items.length}`}
      className={`${label}-card`}
    >
      <Card.Header className={`${label}-card-header`}>
        <OverlayTrigger
          placement="top"
          overlay={
            isExpense ? (
              <Tooltip
                id={`tooltip-value-${label}`}
                style={{ position: "fixed" }}
              >
                {revenuePercentage}% of revenue
              </Tooltip>
            ) : (
              // biome-ignore lint/complexity/noUselessFragments: react types need it
              <></>
            )
          }
        >
          <Row>
            <Col>{label}</Col>
            <Col className="text-end fixed-width-font">
              <div>{intlFormat(roundBig(Big(total), 2), userOptions)}</div>
            </Col>
          </Row>
        </OverlayTrigger>
      </Card.Header>
      <Card.Body>
        {table?.items && (
          <Reorder.Group
            axis="y"
            values={table.items}
            onReorder={reorderTable}
            as="div"
          >
            {table.items?.map((item: BudgetItem) => (
              <Reorder.Item
                key={item.id}
                value={item}
                as="div"
                dragListener={isDraggable}
              >
                <ItemFormGroup
                  key={`${label}-${item.id}-item-form-group`}
                  itemForm={item}
                  label={label}
                  costPercentage={
                    budget
                      ? BudgetItem.percentage(item.value, budget.incomes.total)
                      : 0
                  }
                  inputRef={inputRef}
                  userOptions={userOptions}
                />
              </Reorder.Item>
            ))}
          </Reorder.Group>
        )}
        <div className="mt-3" />
        <div className="d-flex gap-2">
          <OverlayTrigger
            delay={250}
            placement="top"
            overlay={
              <Tooltip
                id={"tooltip-new-itemformgroup"}
                style={{ position: "fixed" }}
              >
                add new item
              </Tooltip>
            }
          >
            <Button
              variant={`${label.toLowerCase()}-plus-button`}
              aria-label={`add item to ${label}`}
              size="sm"
              className="flex-grow-1"
              onClick={() => {
                addItemToTable(table);
                setTimeout(() => {
                  inputRef.current?.focus();
                }, 0);
              }}
              name=""
              value=""
            >
              <BsPlusLg aria-hidden={true} />
            </Button>
          </OverlayTrigger>
          <OverlayTrigger
            delay={250}
            placement="top"
            overlay={
              <Tooltip
                id={"tooltip-reorder-itemformgroup"}
                style={{ position: "fixed" }}
              >
                toggle reordering of items
              </Tooltip>
            }
          >
            <ToggleButton
              id={`toggle-reorder-${label}`}
              aria-label={`reorder items in ${label}`}
              type="checkbox"
              variant="outline-info toggle"
              value={1}
              size="sm"
              onClick={() => setIsDraggable(!isDraggable)}
              checked={isDraggable}
            >
              <BsArrowsVertical aria-hidden={true} />
            </ToggleButton>
          </OverlayTrigger>
        </div>
      </Card.Body>
    </Card>
  );
}
