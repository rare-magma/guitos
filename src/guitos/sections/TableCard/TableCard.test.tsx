import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router";
import { describe, expect, it } from "vitest";
import { setBudgetMock } from "../../../setupTests";
import { BudgetMother } from "../../domain/budget.mother";
import TableCard from "./TableCard";

describe("TableCard", () => {
  const comp = (
    <BrowserRouter>
      <TableCard header={"Expenses"} />
    </BrowserRouter>
  );

  it("matches snapshot", () => {
    render(comp);
    expect(comp).toMatchSnapshot();
  });
  it("renders initial Expenses state", async () => {
    render(comp);
    expect(await screen.findByDisplayValue("expense1")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$10")).toBeInTheDocument();
  });

  it("renders initial Revenue state", async () => {
    render(
      <BrowserRouter>
        <TableCard header={"Revenue"} />
      </BrowserRouter>,
    );
    expect(await screen.findByDisplayValue("income1")).toBeInTheDocument();
    expect(await screen.findByDisplayValue("$100")).toBeInTheDocument();
  });

  it("responds when user changes input", async () => {
    render(comp);
    await userEvent.type(screen.getByDisplayValue("expense1"), "change name");
    expect(screen.getByDisplayValue("expense1change name")).toBeInTheDocument();

    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...BudgetMother.testBudget(),
        expenses: {
          items: [{ id: 1, name: "expense1change name", value: 10 }],
          total: 10,
        },
      },
      false,
    );
    setBudgetMock.mockClear();

    await userEvent.type(screen.getByDisplayValue("$10"), "123");

    expect(screen.getByDisplayValue("$123")).toBeInTheDocument();
    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...BudgetMother.testBudget(),
        expenses: {
          items: [{ id: 1, name: "expense1", value: 123 }],
          total: 123,
        },
        stats: {
          ...BudgetMother.testBudget().stats,
          available: -23,
          withGoal: -33,
        },
      },
      false,
    );
  });

  it("adds new Expense when user clicks adds new item button", async () => {
    render(comp);
    await userEvent.click(
      screen.getByRole("button", { name: "add item to Expenses" }),
    );
    expect(screen.getByDisplayValue("$10")).toBeInTheDocument();
    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...BudgetMother.testBudget(),
        expenses: {
          items: [
            ...BudgetMother.testBudget().expenses.items,
            { id: 2, name: "", value: 0 },
          ],
          total: 10,
        },
      },
      true,
    );
  });

  it("adds new Revenue when user clicks adds new item button", async () => {
    cleanup();
    render(
      <BrowserRouter>
        <TableCard header={"Revenue"} />
      </BrowserRouter>,
    );
    await userEvent.click(
      screen.getByRole("button", { name: "add item to Revenue" }),
    );
    expect(screen.getByDisplayValue("$100")).toBeInTheDocument();
    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...BudgetMother.testBudget(),
        incomes: {
          items: [
            ...BudgetMother.testBudget().incomes.items,
            { id: 3, name: "", value: 0 },
          ],
          total: 100,
        },
      },
      true,
    );
  });

  it("removes item when user clicks delete item button", async () => {
    render(comp);
    await userEvent.click(
      screen.getByRole("button", { name: "delete item 1" }),
    );
    await userEvent.click(
      screen.getByRole("button", { name: "confirm item 1 deletion" }),
    );
    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...BudgetMother.testBudget(),
        expenses: {
          items: [],
          total: 0,
        },
        stats: {
          ...BudgetMother.testBudget().stats,
          available: 100,
          withGoal: 90,
        },
      },
      true,
    );
  });
});
