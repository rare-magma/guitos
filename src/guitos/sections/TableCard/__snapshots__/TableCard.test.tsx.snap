// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableCard > matches snapshot 1`] = `
<TableCard
  header="Expenses"
/>
`;
